### PR TITLE
change upgrade path to 4.14 (from 4.13) and intensify check

### PR DESCRIPTION
**Pull request by Pearl1594**
## Description
Change upgrade path for image store table

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

Test Output:
Before migrate:
```
MariaDB [cloud]> select * from volume_store_ref\G
*************************** 1. row ***************************
                  id: 8
            store_id: 2
           volume_id: 3
             zone_id: 0
             created: 2020-09-17 10:49:23
        last_updated: NULL
              job_id: NULL
        download_pct: 0
                size: 52428800
       physical_size: 0
      download_state: NULL
            checksum: NULL
           error_str: NULL
          local_path: NULL
        install_path: volumes/2/3/365df8ed-b414-4311-af46-82043ee056af.qcow2
                 url: NULL
        download_url: NULL
               state: Ready
...
*************************** 5. row ***************************
                  id: 22
            store_id: 2
         template_id: 202
             created: 2020-09-17 10:49:26
        last_updated: NULL
              job_id: NULL
        download_pct: 100
                size: 52428800
          store_role: Image
       physical_size: 43646976
      download_state: DOWNLOADED
           error_str: NULL
          local_path: NULL
        install_path: template/tmpl/2/202/6c3ec8aa-77fb-3cac-a74e-70f715361895.qcow2
                 url: NULL
               state: Ready
....
MariaDB [cloud]> select * from snapshot_store_ref\G
************************** 2. row ***************************
                id: 8
          store_id: 2
       snapshot_id: 1
           created: 2020-09-17 10:49:23
      last_updated: NULL
            job_id: NULL
        store_role: Image
              size: 52428800
     physical_size: 43646976
parent_snapshot_id: 0
      install_path: snapshots/2/3/1fae4aa3-270e-490a-9592-3fc3cc40b35a
             state: Ready
      update_count: 2
           ref_cnt: 0
           updated: 2020-09-17 10:49:29
         volume_id: 3

```
After migration:

```
MariaDB [cloud]> select * from volume_store_ref\G
*************************** 1. row ***************************
                  id: 9
            store_id: 1
           volume_id: 3
             zone_id: 0
             created: 2020-09-17 10:53:44
        last_updated: NULL
              job_id: NULL
        download_pct: 0
                size: 52428800
       physical_size: 0
      download_state: NULL
            checksum: NULL
           error_str: NULL
          local_path: NULL
        install_path: volumes/2/3/365df8ed-b414-4311-af46-82043ee056af.qcow2
                 url: NULL
        download_url: NULL
               state: Ready

MariaDB [cloud]> select * from template_store_ref\G
*************************** 6. row ***************************
                  id: 25
            store_id: 1
         template_id: 202
             created: 2020-09-17 10:53:51
        last_updated: NULL
              job_id: NULL
        download_pct: 100
                size: 52428800
          store_role: Image
       physical_size: 43646976
      download_state: DOWNLOADED
           error_str: NULL
          local_path: NULL
        install_path: template/tmpl/2/202/6c3ec8aa-77fb-3cac-a74e-70f715361895.qcow2
                 url: NULL
               state: Ready

MariaDB [cloud]> select * from snapshot_store_ref\G
*************************** 2. row ***************************
                id: 9
          store_id: 1
       snapshot_id: 1
           created: 2020-09-17 10:53:44
      last_updated: NULL
            job_id: NULL
        store_role: Image
              size: 52428800
     physical_size: 43646976
parent_snapshot_id: 0
      install_path: snapshots/2/3/1fae4aa3-270e-490a-9592-3fc3cc40b35a
             state: Ready
      update_count: 2
           ref_cnt: 0
           updated: 2020-09-17 10:53:51
         volume_id: 3

```

### CODE REVIEW

**Comment by Pearl1594** [comment]
@blueorangutan package

**Review by rhtyd — APPROVED** [review]

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan test

**Review by rhtyd — APPROVED** [review]

File: engine/schema/src/main/resources/META-INF/db/schema-41310to41400.sql [hunk 1]
```diff
@@ -53,33 +53,6 @@ ALTER TABLE `cloud`.`vm_instance` ADD COLUMN `backup_offering_id` bigint unsigne
 ALTER TABLE `cloud`.`vm_instance` ADD COLUMN `backup_external_id` varchar(255) DEFAULT NULL COMMENT 'ID of external backup job or container if any';
 ALTER TABLE `cloud`.`vm_instance` ADD COLUMN `backup_volumes` text DEFAULT NULL COMMENT 'details of backedup volumes';
 
-ALTER TABLE `cloud`.`image_store` ADD COLUMN `readonly` boolean DEFAULT false COMMENT 'defines status of image store';
```
Comment (DaanHoogland) [inline engine/schema/src/main/resources/META-INF/db/schema-41310to41400.sql:56]: why is this moved to another upgrade script? it makes it harder to compare and would imply it will not be needed/used in propriatory 4.13 releases anymore

File: services/secondary-storage/server/src/main/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java [hunk 2]
```diff
@@ -1049,8 +1049,11 @@ protected Answer execute(CopyCommand cmd) {
         DataTO destData = cmd.getDestTO();
         DataStoreTO srcDataStore = srcData.getDataStore();
         DataStoreTO destDataStore = destData.getDataStore();
-
-        if (DataStoreRole.Image == srcDataStore.getRole()  && DataStoreRole.Image == destDataStore.getRole()) {
+        if (DataStoreRole.Image == srcDataStore.getRole() && DataStoreRole.Image == destDataStore.getRole() &&
+            srcDataStore instanceof NfsTO && destDataStore instanceof NfsTO &&
+            ((srcData.getObjectType() == DataObjectType.TEMPLATE && destData.getObjectType() == DataObjectType.TEMPLATE) ||
+                    (srcData.getObjectType() == DataObjectType.SNAPSHOT && destData.getObjectType() == DataObjectType.SNAPSHOT) ||
+                    (srcData.getObjectType() == DataObjectType.VOLUME && destData.getObjectType() == DataObjectType.VOLUME))) {
```
Comment (DaanHoogland) [inline services/secondary-storage/server/src/main/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java:1056]: please put this condition in a boolean method with a good descriptive name.

**Comment by harikrishna-patnala** [comment]
LGTM

**Review by harikrishna-patnala — APPROVED** [review]

**Review by DaanHoogland — CHANGES_REQUESTED** [review]

File: engine/schema/src/main/resources/META-INF/db/schema-41310to41400.sql (hunk 1 above)
Comment (harikrishna-patnala) [reply engine/schema/src/main/resources/META-INF/db/schema-41310to41400.sql:56]: @DaanHoogland  This is related to PR#4053 which merged recently to master where the schema changes were made in old schema file. 
https://github.com/apache/cloudstack/pull/4053/files#r490066335

**Review by shwstppr — APPROVED** [review]
LGTM

File: services/secondary-storage/server/src/main/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java (hunk 2 above)
Comment (Pearl1594) [reply services/secondary-storage/server/src/main/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java]: @DaanHoogland extracted the conditional

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

**Review by DaanHoogland — APPROVED** [review]

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan test

**Review by rhtyd — APPROVED** [review]
Tests lgtm

**Comment by rhtyd** [comment]
Manually verified, no regressions seen. The tests finished at backend
